### DISCUSSION TITLE
Event state and multiplicity of events referenced

### DISCUSSION THREAD

**Discussion by mswiderski** [serverlessworkflow/specification#581] [General]
While looking at event state it defines `onEvents` which is a list of event actions that should happen plus it references an array of event definitions it should react to. So from user perspective it seems like two levels

- `onEvents` sort of means  events to be listening for and perform actions on receive
- `eventRefs` are actual references to the events to listen for

The first problem I see is that the `exclusive` attribute is on the same as `onEvents` which in my opinion is correct as then it would apply to all items of the `onEvents` array. And by that I'd say `OnEvent` should have only one reference to event definition and not array. 

From runtime implementation point of view event definitions can be different from following perspectives
- data format it comes with thus same actions won't be able to properly filter data out of events
- different integration like Kafka topic, MQTT topic, HTTP endpoint etc that would make it really hard and complex to have support like that

With these point in mind I'd like to suggest to change the `OnEvents` definition to have only one `eventRef` instead of the list as this will make it easier to consume from users of the workflow definition. Plus it will be really explicit how the `exclusive` attribute applies. As currently it is not rally known what should happen. Let's take an example

- there are 5 `onEvents` array items defined
- each of the items reference 3 event definitions
- `exclusive` attribute is set to `false`

would this setup mean that it will only continue if there will be 15 events received?

Changing to what I suggest it would be clear that having
- there are 5 `onEvents` array items defined
- each of the items reference 1 event definition
- `exclusive` attribute is set to `false`

it will only continue when 5 events are received and that each event definition can actually come from distinct channels and have their own specific data formats.

What are your thoughts guys?

**Comment by mswiderski** [comment]
Another alternative could be to make `onEvents` to be `onEvent` (one item instead of array) and then move the exclusive attribute to the `OnEvent` definition that would then apply to `eventRefs` and allow to define clearly the logic around handling events.

**Comment by tsurdilo** [comment]
@mswiderski 

        The first problem I see is that the exclusive attribute is on the same as onEvents which in my opinion is correct as then it would apply to all items of the onEvents array. And by that I'd say OnEvent should have only one reference to event definition and not array.

I think there is a use case for having eventRefs as array and it has to do with correlation. Currently event definitions are domain-specific and referenced by name only. There is nothing to prevent me from defining:

      "events": [
      {
            "name": "MyEventA",
            "type": "my.events",
            "source": "/myevents",
            "correlation": [
              {
                "contextAttributeName": "someattributeA",
                "contextAttributeValue": "abc"
              }
            ]
          },
          {
            "name": "MyEventB",
            "type": "my.events",
            "source": "/myevents",
            "correlation": [
              {
                "contextAttributeName": "someattributeB",
                "contextAttributeValue": "xyz"
              }
            ]
          }
        ]

and then do:

     "eventRefs": ["MyEventA", "MyEventB"]

In this case its same event but different correlation defined. 
Do you think this can work, meaning we say that ok it can be multiple only if its same event (type+source) or in this case should we use expressions to define this correlation instead?

**Reply by mswiderski** [reply]
yes, it makes sense and I think will work. As runtime would simply consume event and try to correlate with existing instances - as correlation essentially means to find instances that should consume it.

For start events runtime can also say that if correlation does not match it can start new instance. With this we cover both scenarios
- start new if not already running 
- deliver event to already running instance

second is particularly important for event state with `exclusive` set to false.

I am leaning more and more towards that there should be single `onEvent` and multiple `eventRefs` and this makes it the most clear if you ask me.

**Reply by fjtirado** [reply]
@mswiderski I think all depends on how much complexity we want to allow the user to put into the flow. 
If we want alternative actions to be executed by the state depending on the event type consumed, then we need multiple `onEvent`. If we want that the same actions are always executed regardless which event type, then single `onEvent` is the way to go. 
But, if we keep multiple `onEvent` (we go for flexibility at the expense of simplicity), I'm pretty convinced we need to move `exclusive` inside every `onEvent` (clearly implying  that `exclusive` applies to the multiple `eventRef` there) and add a new field at `state` level (candidate name can be `waitOnAllEvents`) that indicates whether we should wait for all onEvents to be processed before transitioning to the next state or not. 
Thtat the rationale behinds @ricardozanini PR https://github.com/serverlessworkflow/specification/pull/594

**Reply by cdavernas** [reply]
@fjtirado I completely agree with you here! +100

**Reply by mswiderski** [reply]
that's my point - I'd go for simplicity first until we get demand for complex scenarios. As with single `onEvent` you can still achieve different actions for different events by using different event states. So yes, it will be more verbose but in my opinion way easier to understand. 
Since we (who implement the spec) struggle with clear positioning imagine being in users shoes...

With that said, I'd make it simple and allow only one `onEvent` for event state to start with.

**Reply by fjtirado** [reply]
Ok, if we leave just one single onEvent, can we still move "exclusive" inside of the onEvent definition?. If later, we decide to add more onEvent, we will have the flexibility to do that without really doing breaking changes in the spec (because, as I tried to explain, multiple onEvents will require another field and because the exclusive policy might be different for any of them)

**Reply by ricardozanini** [reply]
@mswiderski you replied to the wrong thread I think 😆

**Reply by mswiderski** [reply]
completely ok to move the `exclusive` property into `OnEvent` definition. Makes a lot of sense to keep it together.

**Reply by mswiderski** [reply]
> @mswiderski you replied to the wrong thread I think 😆

not sure why, but it might be the case... quite few things happening so errors can be expected :)

**Reply by tsurdilo** [reply]
think i need to see some Current vs A vs B json/yaml example to be able to say more about all this :) 
hard to follow w/o seeing that

**Reply by ricardozanini** [reply]
@tsurdilo, I think this is what we are looking for (some attributes suppressed to facilitate the comprehension):

```json

{
  "states":[
    {
     "name": "stateA",
      "type": "event",
      "onEvent":{
        "exclusive":true,
        "eventRefs":[
          "HighBodyTemperature",
          "HighBloodPressure"
        ],
        "actions":[
          {
            "functionRef":{
              "refName":"SendTylenolOrder",
              "arguments":{
                "patient":"${ .patientId }"
              }
            }
          },
          {
            "functionRef":{
              "refName":"CallNurse",
              "arguments":{
                "patient":"${ .patientId }"
              }
            }
          }
        ]
      }
    },
    {
      "name": "stateB",
      "type": "event",
      "onEvent":{
        "exclusive":false,
        "eventRefs":[
          "HeartBPDown",
          "CodeBlue"
        ],
        "actions":[
          {
            "functionRef":{
              "refName":"CallSurgeon",
              "arguments":{
                "patient":"${ .patientId }"
              }
            }
          }
        ]
      }
    }
  ]
}
```

The changes are:

1. Make the `onEvents` attribute from the `Event` state *singular*. This way, it simplifies the declaration of the state and what happens once we consume an event.
2. Move the attribute `exclusive` to the newly `onEvent`. This way, we can control more easily the event aggregation, mainly for `exclusive=false`, which is a reasonably complex infrastructure problem. I'll open a separate discussion thread about this topic later.

I can change my PR to reflect this change. @mswiderski @fjtirado @cdavernas cc

**Reply by mswiderski** [reply]
+1 from me on this!

**Reply by tsurdilo** [reply]
My question then is why even have actions on the event state. 
Also is StateA transitioning to StateB or are we thinking to add multiple starting workflow states? This does not exist currently and don't think we want to add. 

 "HeartBPDown",
"CodeBlue"
events in this case are not workflow starting events, is it assumed they are?

How do we solve the use case of
1. "IF Event A and B go route X"
2. "IF Event C or event D go route Y"
without multiple starting workflow states i think 2. will always have to depend on 1. happening first

**Reply by tsurdilo** [reply]
@mswiderski @ricardozanini 
imo a maybe better solution for this is to change eventRefs to eventRef and only allow a single event, and have exclusive on the top like currently. 
it would imo accomplish the same thing, but would allow us not to have to add multiple starting workflow states. wdyt?

bpmn allows multiple starting events and i kinda fail to understand why with this it could not be translated into that

**Reply by fjtirado** [reply]
The problem I foresee with this latter approach is that if you want that your flow starts when events A or B  arrive and you want to execute the same set of actions for both, you need to duplicate your action definition (since `eventRef` will only allow just one event)  in both `onEvents`
I would rather assume exclusive always true (so remove exclusive) and have either multiple `eventRefs` for one `onEvent` (meaning that if A or B arrives the same actions will be executed) or mutiple `onEvents` with one distinct `eventRef` for each of them (meaning that when A arrive some actions are executed and when B arrives other actions are executed). In all cases, the reception of one event means that the flow is started or resumed (which changes, depending on the definition, is the actions that are executed upon event reception, not the moment in which the transition to the next state happen). 
Exclusive set to false is a simplified way of "event correlation", meaning that once an event is received, we have to wait for another one in other to start or resume the flow . And this, in my opinion, is more the work of an external tool than a flow processor. 
Exclusive set to false has certain implications for event timeout.  Lets assume you have one `onEvent` for `eventRef` A and another `onEvent` with `eventRef`  B and `exclusive` set to false. Event A arrives, so event A actions are executed (if you want the actions associated to A to be executed only when A and B are received, then you should use multiple `eventRef` with just one `onEvent`) , we are still waiting for event event B to arrive, but if the timeout ocurrs before this event arrive, according to existing event timeout expiration behaviour, not all actions defined for that event state when exclusive is set to false will be executed before the transition to the  next state occur, which Im not sure is the desired behaviour. This can be probably sorted out by extending event timeout expiration behaviour, but it is a factor to consider in whatever decision we take about this matter.

**Reply by tsurdilo** [reply]
not sure i understand. 
event state as its now has imo most power of what can be expressed but apparently its too hard to implement. dumming it down seems also not wanted. 
still dont know if the sample
json assumes multiple starting stares or not. 
we had all these types of discussions on event state before and i rather not have to relive that again :) that said tho id suggest either making it completely simple and go from there or leaving it as it and live with possibility of misinterpretation until we have more data how ppl use it

**Reply by fjtirado** [reply]
Implementation is not really a concern, luckily ;). Currently, we are already implementing both exclusive false and true (as we intepreted it, that it applies both to `eventRef` and `onEvents`). I was just trying to point out the possible interpretation issues of what we currently have. In fact our first proposal was to keep both `onEvents` and `eventRef`, move exclusive inside  `onEvent` (so it is make clear that it applies to `eventRef`) and define a new field for establishing which is the relation between several `onEvents`, to make it even more flexible. In my opinion, the problem of the current text (which appart from that minor thing is great on defining a flexible and powerful event state) in the spec is that it is unclear the scope of `exclusive` field (if it applies to `eventRef` or `onEvents` or both, and in case it only applies to one of them, what happens with the other construct)

**Reply by tsurdilo** [reply]
yeah the exclusive flag change is something everyone here is asking for so lets do at least that for starters maybe and go from there?

**Reply by ricardozanini** [reply]
Ok, so you want to keep the event matrix for `onEvents` and `eventRefs` to tackle a wider set of use cases? Summarizing the example, we then will have only one state:

```json
{
   "states":[
    {
      "name": "stateA",
      "type": "event",
      "onEvents": [{
        "exclusive":true,
        "eventRefs":[
          "HighBodyTemperature",
          "HighBloodPressure"
        ],
        "actions":[
          {
            "functionRef":{
              "refName":"SendTylenolOrder",
              "arguments":{
                "patient":"${ .patientId }"
              }
            }
          },
          {
            "functionRef":{
              "refName":"CallNurse",
              "arguments":{
                "patient":"${ .patientId }"
              }
            }
          }
        ]
      }, {
        "exclusive":false,
        "eventRefs":[
          "HeartBPDown",
          "CodeBlue"
        ],
        "actions":[
          {
            "functionRef":{
              "refName":"CallSurgeon",
              "arguments":{
                "patient":"${ .patientId }"
              }
            }
          }
        ]
      }
   ]}
```

This will lead to:

1. After receiving `HighBodyTemperature` **or** `HighBloodPressure` actions will be executed, then we transition to the second state; **OR**
2. After receiving `HeartBPDown` **and** `CodeBlue` actions will be executed, then we transition to the second state;

@mswiderski I believe that's simple enough as well, wdyt?

I still don't like the idea of aggregating events based on a correlation key since it feels like it's not the spec responsibility to dictate that. In such a scenario, I would delegate the  `HeartBPDown` and `CodeBlue` events correlation to another system, such as a CEP or Event Mesh to correlate them for me, generate an aggregated event, and then we start the workflow based on that new event. 

The correlation of events can get complex, and soon, people might start asking for more. But let's keep this discussion on a new thread.

**Reply by cdavernas** [reply]
> I still don't like the idea of aggregating events based on a correlation key since it feels like it's not the spec responsibility to dictate that. In such a scenario, I would delegate the HeartBPDown and CodeBlue events correlation to another system, such as a CEP or Event Mesh to correlate them for me, generate an aggregated event, and then we start the workflow based on that new event.

I disagree here. No system that I could put my hands on was able to correctly correlate events, or even do it at all. It is a runtime concern that can easily be implemented with the guidelines set out by the spec, which are basically attribute key/value mappings. 
As for aggregation, it is a whole different concern, which does not apply IMO.

**Reply by ricardozanini** [reply]
@cdavernas let's agree to discuss this in another thread. :)

**Reply by mswiderski** [reply]
> This will lead to:

> After receiving HighBodyTemperature or HighBloodPressure actions will be executed, then we transition to the second state; OR
> After receiving HeartBPDown and CodeBlue actions will be executed, then we transition to the second state;
> @mswiderski I believe that's simple enough as well, wdyt?

I am still in favour of having single `onEvent` and then multiple `eventRefs` as to me this is the most readable setup. Event state react to event(s) defined as part of the event refs and then depending on the `exclusive` value it will either trigger as soon as one even arrives or wait for all of these event and call associated action before transitioning.

In my opinion what is currently is way more than what's needed in real uses cases as a good practice should be that the workflows are really specific to do the job well rather than try to be a generic so it can cover all uses cases possible.

**Reply by ricardozanini** [reply]
I also agree with this simplified approach, I guess we need to convince @tsurdilo now. :) @cdavernas wdyt?

**Reply by fjtirado** [reply]
@mswiderski  Are we supporting multiple start states? Im asking because I believe  `start` field in the flow header only supports one state. In other words, If I'm right, if you want to execute different actions depending on the event type at startup, you still need multiple `onEvents` because you cannot specify multiple start states.

**Reply by mswiderski** [reply]
I don’t think there is currently support for multiple start states and I’d keep it that way. Either explicitly by using the `start` attribute or as first state in the array. 

I don’t think executing different actions depending on events is what is required as that in my opinion would be not a common requirement as it could lead to complexity in the workflow. 
And if you need that you can do that as well based on switch state so there is already way of doing that. I am more into having less options of doing the same thing than more as it is easier to consume.

**Reply by fjtirado** [reply]
Yes, I agree with not having multiple start events (that will really complicate the whole thing)
Switch state will work if we add the received event type to the flow data, which can be done if `dataOnly `is set to false, so yes, we can kill the onEvent. 
So basically, the idea is to put all the events that should trigger the flow on the eventRef and has a switch state afterward that selects the actions to be executed based on the eventType (achieving the same purpose than onEvents)
Perfect

**Reply by tsurdilo** [reply]
> I am still in favour of having single onEvent and then multiple eventRefs as to me this is the most readable setup. 

i think in this case we no longer need "onEvent", can simplify by removing it

> So basically, the idea is to put all the events that should trigger the flow on the eventRef and has a switch state afterward that selects the actions to be executed based on the eventType (achieving the same purpose than onEvents)

that is pretty much the reason why we made event state as it is now, to avoid having to mandate switch state right after :) Is funny how things go full circle. I wonder what the benefits are of doing it like this.  
if we go this route where event state no longer needs actions thats fair as well, but then i dont see a difference between that and 
just removing event state and using operation state with an "eventRef" action that specifies only consumeEventRef(s)

**Reply by tsurdilo** [reply]
what im trying to say is that we already have the "start" definition that can be of object type

if we are going in the direction of only using event state for defining events that can trigger workflow invocation, then i dont think it should be a workflow state at all as it defines no control flow logic but instead be defined inside the start property.like cron etc

i think i would actually be ok with that, as then a starting data-based switch workflow state would make more sense.

**Reply by mswiderski** [reply]
I don’t think that removing actions is of any value, it’s there and it has its purpose. The value of event state is that it allows to build advanced wait state for incoming events and that has big value. But only one level is enough to do majority of use cases. 

I said you can use switch but not that you need it always but only if you need to actually take different paths which imo will be rare. As the complexity of using switch vs multiple levels in event state is much easier to follow and represent at runtime - what other was taken etc. 

Keep in mind that event state allows to have multiple events with the exclusive property and that won’t be easily done not with top level property not operation state. Moreover operation is in my opinion outbound and not inbound.

**Reply by tsurdilo** [reply]
if we have a single "onEvent" (single path) as mentioned then imo there is no need for actions as it can be followed by operation state. 
if that single onEvent can have a multiple eventRefs, i think thats pretty easy to define in start definition and not a workflow state, for example:

```
{
  "id": "myeventstartworkflow",
  "version": "1.0",
  "specVersion": "0.8",
  "name": "my workflow",
   "start": {
     "stateName": "MyOperationState",
     "event": {
       "refs": "A, B",
       "exclusive": true 
     }
   },
  "states": [
    {
      "name": "MyOperationState",
      "type": "operation",
      "actions": [
        
      ],
      "end": true
    }
  ]
}
```

**Reply by fjtirado** [reply]
@tsurdilo Since you mention action eventRef, although not related with this discussion, but following the simplicity spirit exposed by @mswiderski, I feel there is certain overlap between consumeEventRef action and Event and Callback states. I mean, I think we are already covering all scenarios with Callback and Event state (even if we remove onEvent from the latter), so, in my opinion, we will simplify the implementation a lot if we remove consumeEventRef from the spec. You can implement fire&wait scenario with a callback state. You can implement wait&fire with event state. wdyt? (appart that this post should be in a different thread ;))

**Reply by tsurdilo** [reply]
for eventRef action, consumeEventRef is needed to denote sync invocation of that action imo.
yeah there are some overlaps between things, but thats imo fine, eventRef action takes 2 lines of json/yaml compared to many more if workflow state is used

**Reply by fjtirado** [reply]
for the case described in the first line (I call it fire&wait event), it can be a Callback state where the action is a produceEvent one. It will be more or less the same number of lines than an OperationState with an produceEvent and a consumeEvent, and for some reason (this is pretty subjetive) I see flow clearer.

**Reply by hirenr** [reply]
So, do we have a conclusion. Its been some time since the last discussion.

**Reply by ricardozanini** [reply]
This is on our backlog to solve for 0.9, @hirenr. We are looking for contributions if you're willing to take this one. We can revamp the discussion. @cdavernas cc

**Comment by ricardozanini** [comment]
@mswiderski +1 to make `onEvent` a single entry rather than an array. For the `exclusive` use case, this can lead to misunderstanding for all the parts involved. I've opened a PR to introduce a more complex use case to this scenario (multiple `onEvents`, with multiple `eventRefs`): https://github.com/serverlessworkflow/specification/pull/594

**Reply by cdavernas** [reply]
@ricardozanini What about @fjtirado suggestion, which is IMO much more consensual: we keep the possibility to define an array of effects specific to one/many events (aka onEvent), but clarify how they be consumed?

**Reply by ricardozanini** [reply]
Yup, that was my first suggestion and the rationale behind the PR I sent yesterday. Although I like simple things and having onEvent not being an array would simplify the understanding. It's a trade off, and I'm happy with both paths. I would go either way. What we can't have is misunderstanding in the spec.

**Reply by cdavernas** [reply]
I totally agree with you, and remember having to come back and forth to TO at the time I implemented it on Synapse, because of lack of clarity. I'd be more than happy to have it written in stone before I get to refactor it, too 😆

**Reply by fjtirado** [reply]
Yes, same here, I have implemented it in three different ways in the last two days ;). Main doubt is the decision to wait for all onEvents or not (and if exclusive field also applies to that). And realizing that some users, when using different `onEvents`, might want to have the flexibility to use different `exclusive` values for the `eventRefs` inside the `onEvents`.

**Comment by cdavernas** [comment]
Closing this discussion as it applies to a discontinued version of the spec.

I invite all participants to take a look at the [new DSL](https://github.com/serverlessworkflow/specification/blob/main/dsl-reference.md#listen) and open a new discussion or issue if any of the discussed problem persist!
